### PR TITLE
Coupons and sales

### LINKED ISSUE #365: Coupon codes
Any thought about implementing a coupon code system into Saleor?

### PR DESCRIPTION
**Pull request by mociepka**
Features:
* [x] one module with all discounts,
* [x] coupons/vouchers functionality,
* [x] discount form in checkout:
    image

Fixes #365

### CODE REVIEW

**Comment by mociepka** [comment]
Finalny ready to merge @patrys 

## example
image

image

File: saleor/checkout/core.py [hunk 1]
```diff
@@ -133,6 +134,50 @@ def billing_address(self, address):
...
+        currency = self.storage.get('discount_currency')
+        name = self.storage.get('discount_name')
+        if value is not None and name is not None and currency is not None:
+            amount = Price(value, currency=currency)
+            return FixedDiscount(amount, name)
+
+    @discount.setter
+    def discount(self, discount):
+        amount = discount.amount
+        self.storage['discount_value'] = str(amount.net)
```
Comment (patrys) [inline saleor/checkout/core.py]: We need to call Django's smart string here so it works in Python 2.

File: saleor/checkout/templates/checkout/snippets/voucher-form.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                    <span class="text-success">{{ discount.name }}</span>
+                    <button type="submit"
+                            class="btn btn-link btn-xs"
+                            aria-label="{% trans "Remove" %}">
+                            {% bootstrap_icon "remove" %}
+                    </button>
+                </form>
+            </td>
+            <td class="noborder text-right">
+                -&nbsp;{% net discount.amount %}
```
Comment (patrys) [inline saleor/checkout/templates/checkout/snippets/voucher-form.html]: We can't just stick a minus in front of the string, we need to pass a negative value to formatter as where minus goes varies between locales.

Comment (mociepka) [inline saleor/checkout/templates/checkout/snippets/voucher-form.html]: Maybe discount should be negative by default?

File: saleor/discount/models.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        else:
+            return discount
+
+    def get_discount_for_checkout(self, checkout):
+        if self.type == Voucher.VALUE_TYPE:
+            cart_total = checkout.cart.get_total()
+            limit = self.limit if self.limit is not None else cart_total
+            if cart_total < limit:
+                msg = pgettext('voucher', 'Discount applicable only for cart '
+                               'total equal or greater than %(amount)s'
```
Comment (patrys) [inline saleor/discount/models.py]: `This offer is only valid for orders over %(amount)s.`

File: saleor/discount/models.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                               '%(currency)s')
+                raise NotApplicable(msg % {
+                    'amount': limit.net,
+                    'currency': limit.currency})
+            return self.get_fixed_discount_for(cart_total)
+
+        elif self.type == Voucher.SHIPPING_TYPE:
+            if not checkout.is_shipping_required:
+                msg = pgettext('voucher', 'Checkout is not shippable')
```
Comment (patrys) [inline saleor/discount/models.py]: `Your order does not require shipping.`

File: saleor/discount/models.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            if (self.apply_to is not None and
+                    shipping_method.country_code != self.apply_to):
+                msg = pgettext(
+                    'voucher', 'Discount apply only for %(country)s')
+                raise NotApplicable(msg % {
+                    'country': self.get_apply_to_display()})
+            if self.limit is not None and shipping_method.price > self.limit:
+                msg = pgettext(
+                    'voucher', 'Discount applicable only for shipping price '
+                               'equal or greater than %(amount)s%(currency)s')
```
Comment (patrys) [inline saleor/discount/models.py]: `This offer is only valid for shipping over %(amount)s.`

File: saleor/discount/models.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                prices = list(
+                    (item[1] for item in get_product_variants_and_prices(
+                        checkout.cart, self.product)))
+            else:
+                prices = list(
+                    (item[1] for item in get_category_variants_and_prices(
+                        checkout.cart, self.category)))
+            if len(prices) == 0:
+                msg = pgettext(
+                    'voucher', 'Any product in cart matches discount')
```
Comment (patrys) [inline saleor/discount/models.py]: `This offer is only valid for selected items.`

File: saleor/discount/models.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                msg = pgettext('voucher', 'Checkout is not shippable')
+                raise NotApplicable(msg)
+            shipping_method = checkout.shipping_method
+            if not shipping_method:
+                msg = pgettext('voucher', 'Select shipping method first')
+                raise NotApplicable(msg)
+            if (self.apply_to is not None and
+                    shipping_method.country_code != self.apply_to):
+                msg = pgettext(
+                    'voucher', 'Discount apply only for %(country)s')
```
Comment (patrys) [inline saleor/discount/models.py]: `This offer is only valid in %(country)s.`

**Comment by patrys** [comment]
:confetti_ball: :fireworks:

**Pull request merged**